### PR TITLE
Clean up Mapper and friends

### PR DESCRIPTION
**Pull request by lbargaoanu**
Extracted `IMapperBase` for consistency. That causes some problems with dynamic dispatch in net461, hence the test changes. We pass the context ourselves, so I hid those overloads. We could simply drop `ResolutionContext.Mapper` and have `ResolutionContext.Map` methods, but that would be more impactful. I guess we could also have both.

### CODE REVIEW

File: src/AutoMapper/ResolutionContext.cs [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
-    public class ResolutionContext : IRuntimeMapper
+    public class ResolutionContext : IInternalRuntimeMapper
     {
         private Dictionary<ContextCacheKey, object> _instanceCache;
         private Dictionary<TypePair, int> _typeDepth;
-        private readonly IRuntimeMapper _inner;
+        private readonly IInternalRuntimeMapper _inner;
 
-        public ResolutionContext(IMappingOperationOptions options, IRuntimeMapper mapper)
+        internal ResolutionContext(IMappingOperationOptions options, IInternalRuntimeMapper mapper)
```
Comment (lbargaoanu) [inline src/AutoMapper/ResolutionContext.cs:15]: If we must, we can keep this public and make `IInternalRuntimeMapper` public.

File: docs/Lists-and-arrays.md [hunk 2]
```diff
@@ -60,6 +60,7 @@ var configuration = new MapperConfiguration(cfg => {
     cfg.CreateMap<Source, Destination>();
 });
 ```
+The setting can be applied globally and can be overridden per profile and per member with `AllowNull` and `DoNotAllowNull`.
```
Comment (lbargaoanu) [inline docs/Lists-and-arrays.md:63]: Perhaps I should make a property `AllowNull`.

**Comment by lbargaoanu** [comment]
@jbogard Review?

File: src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs [hunk 3]
```diff
@@ -268,45 +268,14 @@ private ExpressionResolutionResult ResolveExpression(IMemberMap propertyMap, Typ
                 return null;
             }
             result = matchingExpressionConverter.GetExpressionResolutionResult(result, propertyMap, letPropertyMaps);
-            if(propertyMap.NullSubstitute != null && result.Type.IsNullableType())
+            if(propertyMap.NullSubstitute != null && (result.Type.IsNullableType() || result.Type == typeof(string)))
```
Comment (lbargaoanu) [inline src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs:271]: Perhaps remove the type check altogether? And simply not use `NullSubstitute` when it doesn't make sense.

File: src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs [hunk 4]
```diff
@@ -268,45 +268,12 @@ private ExpressionResolutionResult ResolveExpression(IMemberMap propertyMap, Typ
                 return null;
             }
             result = matchingExpressionConverter.GetExpressionResolutionResult(result, propertyMap, letPropertyMaps);
-            if(propertyMap.NullSubstitute != null && result.Type.IsNullableType())
+            if(propertyMap.NullSubstitute != null && result.ResolutionExpression is MemberExpression && (result.Type.IsNullableType() || result.Type == typeof(string)))
```
Comment (lbargaoanu) [inline src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs:271]: Perhaps remove the type check altogether? And simply not use NullSubstitute when it doesn't make sense.

**Comment by lbargaoanu** [comment]
@jbogard Ping :)

File: src/AutoMapper/ResolutionContext.cs (hunk 1 above)
Comment (jbogard) [reply src/AutoMapper/ResolutionContext.cs:16]: Meh.

**Pull request merged**